src/mod_last_rdbms.erl

```diff
@@ -25,52 +25,66 @@
          remove_user/2]).
 
 -spec init(jid:server(), list()) -> ok.
-init(_Host, _Opts) ->
+init(Host, _Opts) ->
+    prepare_queries(Host),
     ok.
 
+%% Prepared query functions
+prepare_queries(Host) ->
+    mongoose_rdbms:prepare(last_select, last, [username],
+                           <<"SELECT seconds, state FROM last WHERE username=?">>),
+    mongoose_rdbms:prepare(last_count_active, last, [seconds],
+                           <<"SELECT COUNT(*) FROM last WHERE seconds > ?">>),
+    mongoose_rdbms:prepare(last_delete, last, [username],
+                           <<"delete from last where username=?">>),
+    rdbms_queries:prepare_upsert(Host, last_upsert, last,
+                                 [<<"username">>, <<"seconds">>, <<"state">>],
+                                 [<<"seconds">>, <<"state">>],
+                                 [<<"username">>]).
+
+execute_get_last(LServer, LUser) ->
+    mongoose_rdbms:execute_successfully(LServer, last_select, [LUser]).
+
+execute_count_active_users(LServer, Seconds) ->
+    mongoose_rdbms:execute_successfully(LServer, last_count_active, [Seconds]).
+
+execute_remove_user(LServer, LUser) ->
+    mongoose_rdbms:execute_successfully(LServer, last_delete, [LUser]).
+
+execute_upsert_last(Host, LUser, Seconds, State) ->
+    InsertParams = [LUser, Seconds, State],
+    UpdateParams = [Seconds, State],
+    UniqueKeyValues = [LUser],
+    rdbms_queries:execute_upsert(Host, last_upsert, InsertParams, UpdateParams, UniqueKeyValues).
+
+%% API functions
 -spec get_last(jid:luser(), jid:lserver()) ->
-    {ok, non_neg_integer(), binary()} | {error, term()} | not_found.
+    {ok, non_neg_integer(), binary()} | not_found.
 get_last(LUser, LServer) ->
-    Username = mongoose_rdbms:escape_string(LUser),
-    case catch rdbms_queries:get_last(LServer, Username) of
-        {selected, []} ->
-            not_found;
-        {selected, [{STimeStamp, Status}]} ->
-            case catch mongoose_rdbms:result_to_integer(STimeStamp) of
```

Comment:
Could you comment on why we are not returning the errors anymore? For me it looks like this code was too defensive and tried to protect against faulty data, which should not be present in the DB. Is my assumption correct?

Reply:
`mongoose_rdbms:result_to_integer(STimeStamp)` would never fail, we don't allow NULLs for this field.

```sql
CREATE TABLE last (
    username varchar(250) PRIMARY KEY,
    seconds int NOT NULL,
    state text NOT NULL
) CHARACTER SET utf8mb4
```

We use `execute_successfully`, so we need to handle only two cases: `{selected, []}` and `{selected, [{Timestamp, Status}]`, which `decode_last_result` function does.